src/lua_vm/lua_vm.cc

```diff
@@ -0,0 +1,678 @@
+/*
+ * Lua Bindings for RADOS Object Class
+ */
+#include <errno.h>
+#include <setjmp.h>
+#include <string>
+#include <sstream>
+#include <lua.hpp>
+#include "include/types.h"
+#include "json_spirit/json_spirit.h"
+#include "lua_vm/cls_lua_ops.h"
+#include "lua_vm.h"
+
+cls_handle_t h_class;
+cls_method_handle_t h_eval_json;
+cls_method_handle_t h_eval_bufferlist;
+
+/*
+ * Jump point for recovering from Lua panic.
+ */
+static jmp_buf cls_lua_panic_jump;
+
+/*
+ * Handle Lua panic.
+ */
+static int cls_lua_atpanic(lua_State *lua)
+{
+  CLS_ERR("error: Lua panic: %s", lua_tostring(lua, -1));
+  longjmp(cls_lua_panic_jump, 1);
+  return 0;
+}
+
+struct clslua_err {
+  bool error;
+  int ret;
+};
+
+struct clslua_hctx {
+  struct clslua_err error;
+  InputEncoding in_enc;
+  int ret;
+
+  cls_method_context_t *hctx;
```

Comment:
I think it is a great idea to avoid duplicating code in the case that Lua is used in more than one place in Ceph (e.g. objclass and mds balancer). It looks like you are bringing the core of cls_lua here which makes sense, but it seems like this code is still tuned and specific to the objclass environment.

It would be useful to understand what an MDS balancer Lua script needs, in order to figure out what the intersection in a common Lua component would be.